### PR TITLE
Feature/727 implement circuit from workstep 2 and workstep 3

### PR DESCRIPTION
**Pull request by biscuitdey**
# Description

Adds the circuit for workstep2 and workstep3

## Related Issue

#727 #728

## How Has This Been Tested

Manual test by running locally. Automated testing would be included in future PRs.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/identity/bpiAccounts/accounts.module.ts [hunk 1]
```diff
@@ -32,4 +32,5 @@ export const QueryHandlers = [
   ],
   exports: [BpiAccountAgent, BpiAccountStorageAgent],
 })
+// TODO: Move this under a separate folder 'state' instead of 'identity'
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/identity/bpiAccounts/accounts.module.ts:35]: Is this already an open issue?

File: examples/bri-3/src/bri/merkleTree/agents/import MerkleTree from 'merkletreejs';.ts
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/merkleTree/agents/import MerkleTree from 'merkletreejs';.ts:1]: that is an odd file name ... sure about that?

File: examples/bri-3/src/bri/merkleTree/agents/merkleTreeStorage.agent.ts [hunk 2]
```diff
@@ -55,6 +55,9 @@ export class MerkleTreeStorageAgent extends PrismaService {
       where: { id: merkleTree.id },
     });
 
+    // TODO: Mil5 - Mapping from a prisma client type to a domain object is not working.
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/merkleTree/agents/merkleTreeStorage.agent.ts:58]: open issue or will be resolved before PR merged?

File: examples/bri-3/src/bri/state/agents/state.agent.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { Injectable } from '@nestjs/common';
+
+import MerkleTree from 'merkletreejs';
+import { BpiAccountStorageAgent } from '../../identity/bpiAccounts/agents/bpiAccountsStorage.agent';
+import { BpiAccount } from '../../identity/bpiAccounts/models/bpiAccount';
+import { MerkleTreeAgent } from '../../merkleTree/agents/merkleTree.agent';
+import { MerkleTreeStorageAgent } from '../../merkleTree/agents/merkleTreeStorage.agent';
+import { Witness } from '../../zeroKnowledgeProof/models/witness';
+
+// TODO: We should follow this approach everywhere for storage
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/state/agents/state.agent.ts:10]: TODOs are open issues or will be resolved before PR merged?

File: examples/bri-3/src/bri/transactions/agents/transactions.agent.ts [hunk 4]
```diff
@@ -177,6 +177,7 @@ export class TransactionAgent {
   ): Promise<TransactionResult> {
     const txResult = new TransactionResult();
 
+    // TODO: Have the hash algh 'sha256' in a single place in env instead of spread around in code
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/transactions/agents/transactions.agent.ts:180]: agree! open issue yet?

File: examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts [hunk 5]
```diff
@@ -41,10 +46,20 @@ export class ExecuteVsmCycleCommandHandler
...
 
+      const workflow = await this.workflowStorageAgent.getWorkflowById(
+        tx.workflowInstanceId,
+      );
+
       try {
-        const txResult = await this.agent.executeTransaction(tx, workstep!);
-        // TODO: When do we update the nonce on the BpiAccount? // Whenever a transaction is initiated
-        // TODO: #702 Update relevant Bpi Account state with txResult sucess
+        const txResult = await this.txAgent.executeTransaction(tx, workstep!);
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/vsm/capabilites/executeVsmCycle/executeVsmCycleCommand.handler.ts:54]: error handling?

**Review by Therecanbeonlyone1969 — COMMENTED** [review]
Left some minor comments.

File: examples/bri-3/src/bri/identity/bpiAccounts/accounts.module.ts (hunk 1 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/identity/bpiAccounts/accounts.module.ts:35]: @Therecanbeonlyone1969  This code was copied from Ognjen's PR. This question has been addressed in PR #732

File: examples/bri-3/src/bri/merkleTree/agents/import MerkleTree from 'merkletreejs';.ts
Comment (biscuitdey) [reply examples/bri-3/src/bri/merkleTree/agents/import MerkleTree from 'merkletreejs';.ts:1]: @Therecanbeonlyone1969  This code was copied from Ognjen's PR. This question has been addressed in PR #732

File: examples/bri-3/src/bri/merkleTree/agents/merkleTreeStorage.agent.ts (hunk 2 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/merkleTree/agents/merkleTreeStorage.agent.ts:58]: @Therecanbeonlyone1969  This code was copied from Ognjen's PR. This question has been addressed in PR #732

File: examples/bri-3/src/bri/state/agents/state.agent.ts (hunk 3 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/state/agents/state.agent.ts:10]: @Therecanbeonlyone1969  This code was copied from Ognjen's PR. This question has been addressed in PR #732

File: examples/bri-3/src/bri/transactions/agents/transactions.agent.ts (hunk 4 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/transactions/agents/transactions.agent.ts:180]: @Therecanbeonlyone1969  This code was copied from Ognjen's PR. This question has been addressed in PR #732

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/utils/computePublicInputs.ts [hunk 6]
```diff
@@ -13,11 +12,11 @@ export const computeEffectiveEcdsaSigPublicInputs = (
   msgHash: Buffer,
   publicKeyHex: string,
 ) => {
+  const ec = new EC('secp256k1');
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/utils/computePublicInputs.ts:15]: @biscuitdey should we not make this configurable ... let's open an issue for that. PR is good for now.

**Review by Therecanbeonlyone1969 — APPROVED** [review]
LGTM

File: examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/utils/computePublicInputs.ts (hunk 6 above)
Comment (biscuitdey) [reply examples/bri-3/src/bri/zeroKnowledgeProof/services/circuit/snarkjs/utils/computePublicInputs.ts:15]: Will add to the issue for configurations (Milestone 5).

**Review by Kasshern — APPROVED** [review]
Assent

**Pull request merged**